cli/src/commands/duplicate.rs

```diff
@@ -43,12 +42,15 @@ pub(crate) fn cmd_duplicate(
     args: &DuplicateArgs,
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
-    let to_duplicate: IndexSet<Commit> =
-        resolve_multiple_nonempty_revsets(&args.revisions, &workspace_command)?;
-    if to_duplicate
-        .iter()
-        .any(|commit| commit.id() == workspace_command.repo().store().root_commit_id())
-    {
+    let to_duplicate: Vec<CommitId> = {
+        let expression = workspace_command.parse_union_revsets(&args.revisions)?;
```

Comment:
> There's a subtle behavior change that an empty revset is no longer rejected
individually, but I think that's good for "jj duplicate".

Why do you think it's good for `jj duplicate`? Do you think any of the other callers of `resolve_multiple_nonempty_revsets()` (`jj abandon`, `jj git push -r`, `jj run`)  should change to use `parse_union_revsets()`?

I don't really know if I want the error on empty revsets. I don't think I've ever passed an empty revset to one of these commands, either accidentally (where I would have wanted the error) or intentionally (where I would not have wanted the error).

Maybe @ilyagr has opinions.

Reply:
> > There's a subtle behavior change that an empty revset is no longer rejected
> > individually, but I think that's good for "jj duplicate".
> 
> Why do you think it's good for `jj duplicate`?

I don't think the user would want to duplicate exactly N revisions by `jj duplicate a b c` (N = 3). The input to `jj duplicate` is rather a set of revisions, where an empty set is valid.

> Do you think any of the other callers of `resolve_multiple_nonempty_revsets()` (`jj abandon`, `jj git push -r`, `jj run`) should change to use `parse_union_revsets()`?

I believe `jj abandon` (and maybe `jj run`) belong to the same category. I'm not sure about `jj git push -r` because it doesn't actually select a set of revisions to push. I personally don't mind if `jj git push -r 'none()'` doesn't fail, but it might be better to check if each set of revisions matches at least one named branch to push.

BTW, I also think `resolve_multiple_nonempty_revsets_default_single()` shouldn't reject duplicated revisions if `all:` is specified.

> I don't really know if I want the error on empty revsets.

I would also want to change it to exit with 0, with a warning or status message.

Reply:
> I don't think the user would want to duplicate exactly N revisions by `jj duplicate a b c` (N = 3).

This is out of scope of this PR, but it would be nice if `jj duplicate <SOURCES> [--to DEST]` duplicates `<SOURCES>` and rebase them on to the `DEST` preserving the original DAG structure. Something like:
```
jj duplicate roots(<SOURCES>)::heads(<SOURCES>)
jj abandon ~<SOURCES> # in the duplicated set
jj rebase -r <DUPLICATED> -d <DEST>
```

Reply:
> I don't really know if I want the error on empty revsets.

I'm also not sure, but I'm leaning towards it being OK to not have errors for empty revsets. I also can't remember many situations when it mattered.

> BTW, I also think resolve_multiple_nonempty_revsets_default_single() shouldn't reject duplicated revisions if all: is specified.

What would this mean?

> t would be nice if jj duplicate <SOURCES> [--to DEST] duplicates <SOURCES> and rebase them on to the DEST preserving the original DAG structure. 

I would call the flag `--destination`, by analogy with `rebase`. I don't have a strong opinion about how useful it is -- when would you use it? It makes sense that looking up the ids for the duplicated revisions is annoying, but is rebasing really the most common command to do after `duplicate`?

I'd prefer a more universal solution, but I'm not sure what it would be.

I did once make a draft for `jj duplicate --edit` or `jj duplicate --checkout`, where you'd either edit or checkout the duplicated revision (when it's one revision, I had some thoughts about whether to pick a root or a head otherwise but I forget them). This is because I thought `jj new` is a common command after `jj duplicate`. Also, you could potentially `jj duplicate --edit` and then `jj rebase -s @`. The design is a bit awkward, though, because of the question of what to do when duplicating many revisions.

Reply:
> > I don't really know if I want the error on empty revsets.
> 
> I'm also not sure, but I'm leaning towards it being OK to not have errors for empty revsets.

Ok, thanks.

> > BTW, I also think resolve_multiple_nonempty_revsets_default_single() shouldn't reject duplicated revisions if all: is specified.
> 
> What would this mean?

`jj new all:some_set_containing_x all:other_set_containing_x` shouldn't fail because the user wouldn't care about the exact number of the parents with `all:`.

> > t would be nice if jj duplicate  [--to DEST] duplicates  and rebase them on to the DEST preserving the original DAG structure.
> 
> I would call the flag `--destination`, by analogy with `rebase`. I don't have a strong opinion about how useful it is -- when would you use it?

When I use `duplicate`, I often want to resurrect old changes (keeping the original history), so the first thing to do after `duplicate` is `rebase -d main`.

> The design is a bit awkward, though, because of the question of what to do when duplicating many revisions.

Yep, that's why I thought `--to`/`--destination` might make sense. It implies that all duplicated revisions are constructed as a sub graph. `--edit` might merge the heads of the duplicated sub graph. Anyway, it's unrelated to this PR, and is broader topic.

Reply:
Another option for `jj duplicate -d` is to call it `jj rebase -d --keep` as hg does.

Regarding my original question about when to use `resolve_multiple_nonempty_revsets()` and when to use `parse_union_revsets()`, it sounds like we agree that it seems fine to relax `resolve_multiple_nonempty_revsets()` to allow empty revsets, which makes it equivalent to `parse_union_revsets()`, right? Or was there still some difference? Either way, I'm fine with merging this PR as is and figure out if and how to unify the behaviors in a later PR.

Reply:
> Another option for `jj duplicate -d` is to call it `jj rebase -d --keep` as hg does.

Ah, yes, that would work, too.

> Regarding my original question about when to use `resolve_multiple_nonempty_revsets()` and when to use `parse_union_revsets()`, it sounds like we agree that it seems fine to relax `resolve_multiple_nonempty_revsets()` to allow empty revsets, which makes it equivalent to `parse_union_revsets()`, right?

I'll send a follow up. My original motivation was there were too many revset helpers, but it seemed hard to consolidate them.